runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -157,40 +160,38 @@ public int hashCode() {
         final CorfuRecordId id = new CorfuRecordId(0, guidGenerator.nextLong());
 
-        // If we are in a transaction, then we need the commit address of this transaction
-        // to fix up as the txSequence
-        if (TransactionalContext.isInTransaction()) {
-            /**
-             * This is a callback that is placed into the root transaction's context on
-             * the thread local stack which will be invoked right after this transaction
-             * is deemed successful and has obtained a final sequence number to write.
-             */
-            class QueueEntryAddressGetter implements PreCommitListener {
-                private CorfuRecordId recordId;
-                private QueueEntryAddressGetter(CorfuRecordId recordId) {
-                    this.recordId = recordId;
-                }
-
-                /**
-                 * If we are in a transaction, determine the commit address and fix it up in
-                 * the queue entry.
-                 * @param tokenResponse
-                 */
-                @Override
-                public void preCommitCallback(TokenResponse tokenResponse) {
-                    recordId.setTxSequence(tokenResponse.getSequence());
-                    log.trace("preCommitCallback for Queue: " + recordId.toString());
-                }
-            }
-            QueueEntryAddressGetter addressGetter = new QueueEntryAddressGetter(id);
-            log.trace("enqueue: Adding preCommitListener for Queue: " + id.toString());
-            TransactionalContext.getRootContext().addPreCommitListener(addressGetter);
-        }
-
-        corfuTable.put(id, e);
-        return id;
+        /**
+         * This is a callback that is placed into the root transaction's context on
+         * the thread local stack which will be invoked right after this transaction
+         * is deemed successful and has obtained a final sequence number to write.
+         */
+         class QueueEntryAddressGetter implements PreCommitListener {
+             private CorfuRecordId recordId;
+             private QueueEntryAddressGetter(CorfuRecordId recordId) {
+                 this.recordId = recordId;
+             }
+
+             /**
+              * If we are in a transaction, determine the commit address and fix it up in
+              * the queue entry.
+              * @param tokenResponse
+              */
+             @Override
+             public void preCommitCallback(TokenResponse tokenResponse) {
+                 recordId.setTxSequence(tokenResponse.getSequence());
+                 log.trace("preCommitCallback for Queue: " + recordId.toString());
+             }
+         }
+
+         QueueEntryAddressGetter addressGetter = new QueueEntryAddressGetter(id);
+         log.trace("enqueue: Adding preCommitListener for Queue: " + id.toString());
+         TransactionalContext.getRootContext().addPreCommitListener(addressGetter);
```

Comment:
@hisundar we never clean PreCommitListeners? Is it ok?

Reply:
The listener is part of the TransactionalContext which is thread local. After the transaction completes or fails it will be garbage collected.